File: Trees/BinaryTrees/TreeDiameter.cpp
#include <bits/stdc++.h>
using namespace std;

typedef struct TreeNode{
    int val;
    TreeNode* left,*right;

    TreeNode(){
    }

    TreeNode(int val){
        this->val = val;
        left = right = NULL;
    }
}TreeNode;

class Util{
	public:
    static TreeNode* getTree(){
        TreeNode* root = new TreeNode(10);
        root->left = new TreeNode(20);
        root->right = new TreeNode(30);

        root->left->left = new TreeNode(40);
        root->left->right = new TreeNode(50);
        root->left->right->left = new TreeNode(60);
        root->left->right->right = new TreeNode(70);
        return root;
    }

    static int height(TreeNode* root){
        if(root == NULL)    return -1;

        int lh = height(root->left);
        int rh = height(root->right);

        return max(lh,rh) + 1;
    }

    static int getDiameterBruteForce(TreeNode* root){
        if(root == NULL)    return 0;

        int leftDia = getDiameterBruteForce(root->left);
        int rightDia = getDiameterBruteForce(root->right);

        int leftHeight = height(root->left);
        int rightHeight = height(root->right);

        int currentDia = 2 + leftHeight + rightHeight;

        return max({leftDia,rightDia,currentDia});
    }

    static pair<int,int> getDiamterOptimal(TreeNode* root){
        if(root == NULL)    return {0,-1};  // dia,height;

        pair<int,int> ld = getDiamterOptimal(root->left);
        pair<int,int> rd = getDiamterOptimal(root->right);

        int leftHeight = ld.second;
        int rightHeight = rd.second;

        int currentNodeDiameter = 2 + leftHeight + rightHeight;
        int currentNodeHeight = max(leftHeight,rightHeight) + 1;
        
        int finalDiam = max({currentNodeDiameter,ld.first,rd.first});

        return {finalDiam,currentNodeHeight};
    }

};

int main(){
    TreeNode* root = Util::getTree();
	cout<<Util::getDiameterBruteForce(root)<<" "<<Util::getDiamterOptimal(root).first;
    return 0;
}

File: Trees/BinaryTrees/TreeDiameter.java
package Trees.BinaryTrees;

public class TreeDiameter {


    private static int treeDiameterSolutionOne(TreeNode root) {
        if (root == null) return 0;

        int leftDiameter = treeDiameterSolutionOne(root.left);
        int rightDiameter = treeDiameterSolutionOne(root.right);

        int leftSubtreeHeight = TreeOperations.getTreeHeight(root.left);
        int rightSubtreeHeight = TreeOperations.getTreeHeight(root.right);


        return Math.max(leftDiameter, Math.max(rightDiameter, leftSubtreeHeight + rightSubtreeHeight + 2));
    }

    private static int finalDiameter = Integer.MIN_VALUE;

    private static int treeDiameterSolutionTwo(TreeNode root) {
        if (root == null) return 0;

        int left = treeDiameterSolutionTwo(root.left);
        int right = treeDiameterSolutionTwo(root.right);

        finalDiameter = Math.max(finalDiameter, left + right);

        return Math.max(left, right) + 1;
    }

    static class Ans {
        int height;
        int diameter;

        public Ans(int height, int diameter) {
            this.height = height;
            this.diameter = diameter;
        }
    }

    private static Ans treeDiameterSolutionThree(TreeNode root) {
        if (root == null) return new Ans(-1, 0);

        Ans leftDiameter = treeDiameterSolutionThree(root.left);
        Ans rightDiamter = treeDiameterSolutionThree(root.right);

        // calculate for current root node based on it's children's.
        int currentNodeDiameter = leftDiameter.height + rightDiamter.height + 2;
        int currentNodeHeight = Math.max(leftDiameter.height, rightDiamter.height) + 1;
        Ans currentNodeAns = new Ans(currentNodeHeight, currentNodeDiameter);


        return currentNodeAns;
    }

    public static void main(String[] args) {
        TreeNode root = TreeNode.getInstance1();
        int res = treeDiameterSolutionOne(root);
        System.out.println("Diameter of Solution One is : " + res);
        System.out.print("Diameter of Tree is : " + finalDiameter);

//        Ans ans = treeDiameterSolutionThree(root);
//        System.out.println("Solution 3 is : " + ans.diameter);
    }
}
